lib/rubocop/cop/rspec/no_expectation_example.rb

```diff
@@ -61,11 +88,22 @@ def on_block(node)
           return unless regular_or_focused_example?(node)
           return if including_any_expectation?(node)
           return if including_any_skip_example?(node)
+          return if including_any_allowed_pattern?(node)
 
           add_offense(node)
         end
 
         alias on_numblock on_block
+
+        private
+
+        def including_any_allowed_pattern?(node)
+          node.each_descendant(:send) do |send_node|
```

Comment:
Does this to find allowed expectations in iterators and other nested statements?
Wondering how this affects performance. We run this only once on an example node, right? Shouldn't be a problem then.

Reply:
Is nesting worth an additional example in specs?

Reply:
I tried it on a large project at hand, and as shown below, it seems to be slowing down a bit, but does not seem to be a major performance issue.

9481 files inspected, no offenses detected
bundle exec rubocop --only RSpec/NoExpectationExample  17.42s user 11.90s system 106% cpu 27.638 total

9481 files inspected, no offenses detected
bundle exec rubocop --only RSpec/NoExpectationExample  19.12s user 13.86s system 89% cpu 36.810 total

Reply:
In the case of the following codes, the ASTs will be those indicated in the comments below them, respectively.

```ruby
it { expect_something }
# node
# s(:block,
#   s(:send, nil, :it),
#   s(:args),
#   s(:send, nil, :expect_something))    <--- send node

it do
  do_something
  expect_something
  do_something
end
# node
s(:block,
#   s(:send, nil, :it),
#   s(:args),
#   s(:begin,    <--- Nodes are nested
#     s(:send, nil, :do_something),    <--- send node
#     s(:send, nil, :expect_something),    <--- send node
#     s(:send, nil, :do_something)))    <--- send node
```

I was going to check target all of the send_node methods and use regular expressions to see if any are specified in AllowedPatterns.

Reply:
Wondering if the usage of ["`" node pattern descendent matching syntax](https://docs.rubocop.org/rubocop-ast/node_pattern.html#for-descendants) along with [dynamic checks](https://docs.rubocop.org/rubocop-ast/node_pattern.html#to-call-functions) would change this result.
In any case, I don't see this as a blocker, can be a theoretical performance improvement in some follow-up PR, or a ticket.